rfd/0139-gogo-must-go.md

```diff
@@ -0,0 +1,279 @@
+
+- Removing all `gogoproto` extensions ahead of time made the final migration significantly easier.
+
+#### Thanos's `gogo/protobuf` migration
+
+[Relevant blog post](https://giedrius.blog/2022/02/04/things-learned-from-trying-to-migrate-to-protobuf-v2-api-from-gogoprotobuf-so-far/)
+
+This contains some general advice, but nothing particularly applicable in terms of specifics.
+
+#### CrowdStrike's `csproto` library
+
+https://github.com/CrowdStrike/csproto
+
+CrowdStrike used `gogo/protobuf` to overcome various problems with Google's original protobuf
+library and subsequently ran into issues during `gogo/protobuf`'s deprecated. `csproto`
+is a library that elects to use the proper protobuf runtime for the appropriate message.
+As a result, you could have an intermingling of different protobuf libraries without compatibility
+issues.
+
+### Our complications
+
+Teleport has a number of complications tied in with our use of `gogo/protobuf` that will make
+our migration unique. Along with our complications will be listed a number of potential
+mitigation strategies to discuss.
+
+**The primary issue here is our use of `gogo/protobuf` extensions.** We should strive
+to remove these extensions while maintaining backwards compatibility.
+
+#### JSON/YAML (de)serialization (removing `gogoproto.jsontag`)
+
+We rely heavily on `gogo/protobuf`'s custom JSON tags for serialization our messages into JSON.
+Additionally, we use a library called [`ghodss/yaml`](https://github.com/ghodss/yaml) to create YAML,
+which also relies on these JSON tags.
+
+Each resource is serialized into JSON before being pushed into our backend, which means the
+protobuf JSON tags have an impact quite far down our stack. There are a number of potential
+mitigations here with differing levels of impact. The primary goal of any mitigation strategy
+here should be **migrate away from gogoproto.jsontag extensions.**
+
+##### Serialize objects in backend as the gRPC wire format
+
+Serializing the objects in the backend as the gRPC wire format has a number of benefits:
+
+- Seems to remove the need for most unmarshaling functions in `lib/services`. (confirm?)
+- The backend has no need for JSON marshaling/unmarshaling, so JSON changes will not
+  affect the backend.
+- Allows us to remove gogoproto extensions without affecting backend storage.
+- Allows us to utilize fields in the backend more efficiently and store larger objects.
+
+The disadvantages here:
+
+- Manually examining the backend is more difficult as the stored values will no longer be
+  human readable.
+- For support cases we will occasionally need to perform surgery in the backend, which
+  requires manual editing of objects. This will be significantly more difficult with the
+  backend serialized in the wire format.
+
+##### Write custom storage objects for the backend
+
+Rather than use gRPC's wire format or serialized JSON format, we should consider writing
+a storage object format. This has a number of benefits:
```

Comment:
Why not use protojson for this? It's a canonical encoding of protobuf, and I don't see us abandoning protobuf anytime soon. My guess is that if we require a hand-rolled json struct for this we'll end up writing a json struct that's almost field-by-field compatible with the protobuf counterpart and we'll just deserialize-serialize it to convert it anyway.

Reply:
I see a few reasons for doing this:

* By writing our own storage object, we decouple the API representation of objects from our backend.
* We can ensure backwards compatibility with existing backends with no migration step, though migration isn't terrible.
* We can use these objects subsequently for CLI display and maintain the current k8s-like JSON/YAML structure that we've got.

What do you think?